### PR TITLE
[factorization] Accept default parameters for solvers + extras

### PR DESCRIPTION
**Pull request by mesnardo**
1. Allow using default parameters for KSP and AmgX solver:

* Returns error if path of KSP configuration file is incorrect.
* Returns error if path of AmgX configuration file is incorrect.
* If parameters `type` and `config` are not provided in input file
  `config.yaml`, PetIBM will use PETSc KSP with default parameters.
* If user provides `CPU` as the type but not the `config` parameter, PetIBM will
  use PETSc KSP with default parameters.
* If user provides `GPU` as the type but not the `config` parameter, PetIBM will
  use AmgX solver with default parameters.

2. `make copy-examples` is now accessible from PetIBM top build directory.

3. Output PETSc log view into file every nsave
In addition to saving the numerical solution, we also output the PETSc log view
into a file called "<time-step>.log" located in the solution folder.
This file contains PETSc performances, runtimes, time stamp, ...

4. Read and write time value into HDF5 file
The time value is written as an attribute of the pressure dataset into the HDF5 file.
When restarting from a non-zero time-step index (`startStep != 0`), the time value is read from the HDF5 file that contains the solution.
The YAML configuration parameters `tstart` was also added to the applications
IBPM and decoupled IBPM and could be used to provide the stating time value.

### CODE REVIEW

File: applications/decoupledibpm/main.cpp [hunk 1]
```diff
@@ -68,7 +68,7 @@ int main(int argc, char **argv)
     PetscReal dt = config["parameters"]["dt"].as<PetscReal>();
 
     // current time
-    PetscReal t = start * dt;
+    PetscReal t = config["parameters"]["tstart"].as<PetscReal>(start * dt);
```
Comment (piyueh) [inline applications/decoupledibpm/main.cpp]: Given that now we can read `t` from HDF5 file, I suggest here we only declare `t` without initializing it, and initialize it in the following `if (start from beginning) else if (restart)`.

**Comment by piyueh** [comment]
I have a comment here (it's just a thought about another way of implementation, not important). 

One of the desired behavior is that we always want to write the simulation time every time we save an HDF5 file, and we always read the simulation time whenever we read a restarting HDF5 file. So another implementation is to combine `writeTimeHDF5` into `write` and combine `readTimeHDF5` into `readRestartData`. In this way, we don't need these two new functions, and we don't need to call extra functions in our application codes. However, the drawback is that we have to change the signatures of `write`, `writeRestartData`, and `readRestartData`, because we need to pass the variable `t`. Another drawback is that API users lose flexibility to choose not to save simulation times.

It's hard to say which implementation is better or not. Each one has its own advantages and drawbacks. So this comment is just a comment.

File: applications/navierstokes/navierstokes.cpp [hunk 2]
```diff
@@ -473,6 +473,19 @@ PetscErrorCode NavierStokesSolver::write(const std::string &filePath)
...
+    {
+    	PetscViewer viewerLog;
+    	ierr = PetscViewerCreate(PETSC_COMM_WORLD, &viewerLog); CHKERRQ(ierr);
+      ierr = PetscViewerSetType(viewerLog, PETSCVIEWERASCII); CHKERRQ(ierr);
+      ierr = PetscViewerFileSetMode(viewerLog, FILE_MODE_WRITE); CHKERRQ(ierr);
+      ierr = PetscViewerFileSetName(
+        viewerLog, (filePath + ".log").c_str()); CHKERRQ(ierr);
+      ierr = PetscLogView(viewerLog); CHKERRQ(ierr);
+      ierr = PetscViewerDestroy(&viewerLog); CHKERRQ(ierr);
+    }
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply applications/navierstokes/navierstokes.cpp]: Done.

**Comment by mesnardo** [comment]
@piyueh, I agree with you.
I changed the signatures of write and read methods for the application codes so that the time value is always accessible.

Ready for second review.

**Comment by piyueh** [comment]
Seems good to go.

**Pull request merged**